velox/functions/prestosql/types/JsonType.cpp

```diff
@@ -245,15 +312,20 @@ void castToJsonFromArray(
   auto inputArray = input.as<ArrayVector>();
 
   auto elements = inputArray->elements();
-  auto elementsRows =
-      functions::toElementRows(elements->size(), rows, inputArray);
-  AsJson elementsAsJson{context, elements, elementsRows};
+  SelectivityVector elementsRows(elements->size(), false);
+  auto elementsToTopLevelRows =
+      AlignedBuffer::allocate<vector_size_t>(elements->size(), context->pool());
+  getElementRowsAndReverseMapping(
+      elements->size(), rows, inputArray, elementsRows, elementsToTopLevelRows);
+
+  AsJson elementsAsJson{
+      context, elements, elementsRows, elementsToTopLevelRows};
 
   // Estimates an upperbound of the total length of all Json strings for the
   // input according to the length of all elements Json strings and the
   // delimiters to be added.
   size_t elementsStringSize = 0;
-  context->applyToSelectedNoThrow(rows, [&](auto row) {
+  rows.applyToSelected([&](auto row) {
```

Comment:
I might be missing something, but looks like this pattern is common enough that will (or probably already is) cause bugs in other parts of the codebase that rely on `applyToSelectedNoThrow()` over non-primitive types, right?

I wonder if we should come up with something more generic, maybe like a `applyToSelectedArrayNoThrow()` helper functions that takes a lambda and passes both `row` and `elemRow` to it automatically? Thoughts?

Reply:
You're absolutely right! The same issue exists in CastExpr::applyArray() and CastExpr::applyMap (e.g., cast(array(varchar) as array(bigint))) that I plan to fix later. The idea of `applyToSelectedArrayNoThrow()` sounds good. I'll think about it. Thank you for this suggestion!

Reply:
Sounds good. I'm thinking that applyToSelectedNoThrow() is also used in many vectorized functions outside of Casts, so this bug might also be affecting some of them.

Reply:
Ouch...

Reply:
@pedroerp So I've thought about making a `applyToSelectedArrayNoThrow()` for this fix, but there is a difficulty for casting arrays or maps. The problem is that when looping over the element vector(s), we not only translate the SelectivityVector of top-level rows to element rows, but also translate rows of the decoded vector to rows of its base vector. So we need to reverse both transformation to map the row where an exception throws back to the corresponding top-level row. The base-row to decoded-row mapping is a one-to-many that cannot be captured by a simple data structure like Buffer. (The current code instead loops over the decoded-rows and check whether there are exceptions at the corresponding base-rows.) Please let me know if you have any suggestions.

Reply:
Updated by moving the common logic to EvalCtx::applyToSelectedElementsNoThrow() and EvalCtx::applyToSelectedPeeledElementsNoThrow(). @pedroerp